# Analysis for study 2

library(meffil)
options(mc.cores=10)

library(ggplot2)
library(viridis)
library(meffonym)

# load df (dat) that contains all info about the samples like 
# batch variables, DNA quantity, etc
load(file="full_sample_info.Rdata")

# load qc summary and qc objects generated by meffil
load("qc-summary.rda")
load("qc-objects.rda")
# we need to know which samples are sex outliers
sex_outliers <- as.character(qc.summary$sex.check$sample.name)
# load methylation data
norm.beta <- load("norm.beta.rda")

# extract median methylated signal and the proportion of 
# undetected probes from the qc objects + qc summary
m_df <- lapply(qc.objects, `[`, c("sample.name","median.m.signal"))
m_df <- data.frame(matrix(unlist(m_df), nrow=length(m_df), byrow=TRUE))
colnames(m_df) <- c("sample","m_signal")
m_df$m.signal <- as.numeric(as.character(m_df$m_signal))
dat <- merge(dat, m_df, by.x="sample.name",by.y="sample")
dat$prop.badprobes <- qc.summary$sample.detectionp.summary$tab$prop.badprobes

# identify samples with >200ng DNA
high_quality <- dat[dat$dna > 200,]
dim(high_quality)
# 135  29
high_quality_samples <- as.character(high_quality$sample.name)

## calculate mean methylation profile for those with >200ng DNA
beta.high.qual <- data.frame(norm.beta[,high_quality_samples])
beta.high.qual$mean_profile <- rowMeans(beta.high.qual)

# create df of participants with <200ng DNA
samplenames <- as.character(colnames(norm.beta))
low_dna_samples <- samplenames[!samplenames %in% high_quality_samples]
beta.remaining <- data.frame(norm.beta[,low_dna_samples])
dim(beta.remaining)

# Now calculate the absolute difference between each participant 
# with <200ng DNA and the composite high quality profile
abs_diff <- beta.remaining
for(i in 1:ncol(abs_diff)){
  abs_diff[,i] <- abs_diff[,i] - beta.high.qual$mean_profile
}
abs_diff <- abs(abs_diff)
abs_diff_means <- data.frame(colMeans(abs_diff))
abs_diff_means$dna <- dat$dna[match(rownames(abs_diff_means), dat$linker)]


# Now plot the relationship between DNA concentration and 
# the three QC variables
library(cowplot)
library(ggpubr)

badprobes <- ggplot(dat, aes(x=dna, y=prop.badprobes)) +
  geom_point(alpha=0.2, color="#440154FF")+
  theme_bw()+
  labs(x = "Total input DNA (ng)", y = "Proportion of undetected probes")

m_signal <- ggplot(dat, aes(x=dna, y=m.signal))+
  geom_point(color="#440154FF", alpha=0.2)+
  theme_bw()+
  labs(x = "Total input DNA (ng)", y = "Median methylated signal")

MAD <- ggplot(abs_diff_means, aes(x=dna, y=colMeans.abs_diff.))+
  geom_point(color="#440154FF", alpha=0.2)+
  theme_bw()+
  theme(plot.title = element_text(size=10))+
  labs(x = "Total input DNA (ng)", y = "Mean absolute difference")

jpeg(filename = "dilution_summary_composite.jpg",width = 10, height = 3, units = "in", res = 600)
ggarrange(badprobes, m_signal, MAD, 
          labels = c("A", "B", "C"),
          ncol = 3, nrow = 1)
dev.off()

# and test the strength of the relationships:
cor.test(dat$dna,dat$prop.badprobes)
cor.test(dat$dna,dat$m.signal)
cor.test(abs_diff_means$dna,abs_diff_means$colMeans.abs_diff.)


### variance plot

# load hannum and horvath clocks
load(file="hannum_horvath_clocks.Rdata")
# load smoking and age df
mbms <- read.csv(file="pheno.Rdata"),header = T)
# smoke_3cat is a categorical variable with three values:
# current, former and never smokers

# reduce mbms to individuals with dnam data
mbms <- merge(mbms, clocks, by.x="sentrix_id",by.y="row.names")
mbms$dna <- dat$dna[match(mbms$sentrix_id,dat$sample.name)]

# add cg05575921 to df, as AHRR variable
norm.t <- as.data.frame(t(norm.beta))
mbms$ahrr <- norm.t$cg05575921[match(mbms$sentrix_id,rownames(norm.t))]

# we want to add in batch effects
mbms$plate <- dat$plate[match(mbms$sentrix_id,dat$sample.name)]

# look at variance
# This essentially runs an ewas of variance with DNA concentration
# - is variance associated with DNA concentration at any sites?
# this section takes a few hours to run

#library(quantreg)
meth <- norm.beta
dim(meth)
stopifnot(colnames(meth)==rownames(dat))
meth.t <- t(meth)
meth.m <- as.matrix(meth)

# we will want to remove associations with cell counts
# (we will also remove variance associated with other factors
# as detailed in the formula below)
cellcounts<-meffil.estimate.cell.counts.from.betas(meth.m,cell.type.reference="blood gse35069 complete",verbose=T)
mbms_for_variance <- merge(dat,cellcounts,by="row.names")

# calculate associations between variance at each cpg and DNA input
variance_results <- list()
for(i in 1:ncol(meth.t)){
  cpg <- meth.t[,i]
  d <- abs(rq(cpg~dna + AgeYRS + bmi + slide + smoke_3cat + gender, data = mbms_for_variance)$resid)
  fit <- lm(d~new_dna,data = mbms_for_variance)
  stats <- coef(summary(fit))["dna",]
  variance_results[[i]] <- stats
}

names(variance_results) <- rownames(meth)
variance_df <- do.call(rbind.data.frame, variance_results)
names(variance_df) <- names(variance_results[[1]])
rownames(variance_df) <- rownames(meth)
save(variance_df,file=paste0(output.dir,"/variance_df_new.Rdata"))

# now prepare to plot the results
probe_details <- meffil.featureset("epic")
variance_for_manhattan <- merge(variance_df,probe_details,by.x="row.names",by.y="name")
colnames(variance_for_manhattan)[5] <- "pval"
variance_for_manhattan$log10p <- -log(variance_for_manhattan$pval,10)
variance_for_manhattan$chromosome <- as.factor(variance_for_manhattan$chromosome)
variance_for_manhattan$chromosome <- ordered(variance_for_manhattan$chromosome,
                                             levels=c("chr1", "chr2","chr3","chr4","chr5",
                                                      "chr6","chr7","chr8","chr9","chr10",
                                                      "chr11","chr12","chr13","chr14","chr15",
                                                      "chr16","chr17","chr18","chr19","chr20",
                                                      "chr21","chr22","chrX","chrY"))

library(viridis)
# plot manhattan
jpeg(filename = "variance_manhattan_dnaconc.jpg",width = 10, height = 4, units = "in", res = 600)
ggplot(variance_for_manhattan, aes(x=position, y=log10p, colour = factor(chromosome))) +
  geom_point() +
  theme_bw()+
  facet_grid(. ~ chromosome, space="free_x", scales="free_x",switch="x") +
  scale_colour_viridis(discrete = T)+
  theme(strip.text.x = element_text(angle = 90)) +
  theme(panel.border=element_blank(),
        strip.background = element_rect(colour="white", fill="gray93"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.spacing = unit(0.1, "lines"))+
  guides(colour=FALSE) +
  labs(x="Position",
       y="-log10 pvalue") +             
  geom_hline(yintercept=-log(0.05/nrow(variance_for_manhattan),10), colour="red") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  ggtitle("Variance manhattan for total input DNA")
dev.off()

# pull up those passing genome-wide threshold
threshold_sites <- variance_for_manhattan[variance_for_manhattan$pval < 0.05/nrow(variance_for_manhattan),]
dim(threshold_sites)
# 17 sites
save(threshold_sites,file="threshold_sites_df.Rdata")
threshold_sites <- threshold_sites[order(threshold_sites$pval),]
write.csv(threshold_sites, file = "threshold_sites.csv",row.names = F, quote = F)

variance_for_manhattan <- merge(variance_df,probe_details,by.x="row.names",by.y="name")
colnames(variance_for_manhattan)[5] <- "pval"
variance_for_manhattan <- variance_for_manhattan[order(variance_for_manhattan$pval),]
write.csv(variance_for_manhattan, file = "variance_ewas.csv",row.names = F, quote = F)

# now test whether lower DNA input obscures the relationship
# between DNA methylation and certain phenotypes
# for this analysis we remove the sex outliers 

sex_outliers <- as.character(qc.summary$sex.check$sample.name)
dat$sex_mismatch <- "no"
dat$sex_mismatch[dat$sentrix_id %in% sex_outliers] <- "yes"
table(dat$sex_mismatch)
mbms_nomismatch <- dat[dat$sex_mismatch=="no",]
dim(mbms_nomismatch)
meth <- meth[,colnames(meth) %in% mbms_nomismatch$sentrix_id]

# age difference (how accurate is the age estimate)

# calculate abs(methylation age - actual age)
mbms_nomismatch$age_diff_horvath <- mbms_nomismatch$horvath.score - mbms_nomismatch$AgeYRS
mbms_nomismatch$age_diff_horvath_abs <- abs(mbms_nomismatch$age_diff_horvath)
mbms_nomismatch$age_diff_hannum <- mbms_nomismatch$hannum.score - mbms_nomismatch$AgeYRS
mbms_nomismatch$age_diff_hannum_abs <- abs(mbms_nomismatch$age_diff_hannum)

stopifnot(colnames(meth) == rownames(mbms_nomismatch))
meth.m <- as.matrix(meth)
fit <- lm(ahrr ~ smoke_3cat,data = mbms_nomismatch)
print(summary(fit))
resid <- residuals(fit)
print("dim resid:")
dim(resid)
length(resid)
cor.test(abs(resid), mbms_nomismatch$new_dna)

mbms_nomismatch$resid <- resid

# plot the age difference for the non-mismatched individuals
agediff_horvath <- ggplot(mbms_nomismatch, aes(dna, age_diff_horvath_abs)) + 
  geom_point(colour="#440154FF", alpha=0.2) + 
  stat_smooth(method = "loess") +
  xlab("Total input DNA (ng)")+
  ylab("abs(age - epigenetic age)")+
  theme_bw()+
  ggtitle("Horvath clock")
agediff_hannum <- ggplot(mbms_nomismatch, aes(dna, age_diff_hannum_abs)) + 
  geom_point(color="#440154FF", alpha=0.2) + 
  stat_smooth(method = "loess") +
  xlab("Total input DNA (ng)")+
  ylab("abs(age - epigenetic age)")+
  theme_bw()+
  ggtitle("Hannum clock")
ahrr <- ggplot(mbms_nomismatch, aes(x=dna, y=abs(resid)))+
  geom_point(color="#287D8EFF", alpha=0.2)+
  theme_bw()+
  #theme(plot.title = element_text(size=10))+
  stat_smooth(method = "loess", se=T) +
  labs(title=paste0("Smoking status"), x = "Total DNA input (ng)", y = "abs(cg05575921 residuals)")

jpeg(filename = "figure_5.jpg",width = 9, height = 3, units = "in", res = 600)
ggarrange(agediff_horvath, agediff_hannum, ahrr,
          labels = c("A", "B","C"),
          ncol = 3, nrow = 1)
dev.off()
